calib/ellcal.c

```diff
@@ -0,0 +1,255 @@
+/*
+ * Phoenix-Pilot
+ *
+ * Ellipsoidal calibration common functions
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-Pilot software
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <math.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/time.h>
+
+#include <libsensors.h>
+
+#include <sensc.h>
+#include <vec.h>
+#include <matrix.h>
+
+
+#define ELLCAL_TAG        "ellcal"
+#define LMA_JACOBIAN_STEP 0.0001
+
+
+/*
+* Target function has a form  of: || S(X - H) || = 1 where:
+*
+* X -> 3x1 matrix of (x,y,z) uncalibrated measurement
+* S -> 3x3 ellipsoid deformation matrix of form:
+*    | p0, p1, p2 |
+*    | p3, p4, p5 |
+*    | p6, p7, p8 |
+* H -> 3x1 ellipsoid center offset matrix of form: (p9, p10, p11)^T
+*
+* We want this function to transform data from some ellipsoid into a unit radius sphere.
+* This function does not check any conic function equalities for ellipsoid.
+*
+* Works well if data is placed around (0,0,0) +- (1,1,1) and the biggest ellipsoid semiaxis is of length ~1.
+*/
+int ellcal_lmaResiduum(const matrix_t *P, const matrix_t *V, float *res, bool log)
+{
+	const matrix_t S = { .data = &P->data[0], .rows = 3, .cols = 3, .transposed = 0 };
+	const matrix_t H = { .data = &P->data[9], .rows = 3, .cols = 1, .transposed = 0 };
+
+	int i;
+	float r = 0;
+	float dataX[3], dataTmpX[3];
+	matrix_t X = { .data = dataX, .rows = 3, .cols = 1, .transposed = 0 };
+	matrix_t tmpX = { .data = dataTmpX, .rows = 3, .cols = 1, .transposed = 0 };
+
+	for (i = 0; i < 3; i++) {
+		MATRIX_DATA(&X, i, 0) = MATRIX_DATA(V, 0, i); /* Copy V into X matrix */
+	}
+
+	matrix_sub(&X, &H, NULL);   /* apply offset */
+	matrix_prod(&S, &X, &tmpX); /* apply ellipsoid deformation matrix */
+
+	/* calculate length of produced vector */
+	for (i = 0; i < 3; i++) {
+		r += MATRIX_DATA(&tmpX, i, 0) * MATRIX_DATA(&tmpX, i, 0);
+	}
+
+	/* Length difference between transformed vector and a unit sphere radius is the residuum */
+	*res = (float)(sqrt(r) - 1);
+
+	return 0;
+}
+
+
+/*
+* Jacobian is calculated numerically by calculating differences of residuum after step in parameters.
+* Steps are taken for each parameter at a time.
+*/
+int ellcal_lmaJacobian(const matrix_t *P, const matrix_t *V, matrix_t *J, bool log)
+{
+	float dataPtmp[12] = { 0 };
```

Comment:
Unnecessary zeroing.
```suggestion
	float dataPtmp[12];
```